### PR TITLE
[WIP] Fast tokenizer for debertaV2

### LINKED ISSUE #11529: Deberta v2 Fast Tokenizer
Fast tokenizers for deberta models were requested in #10498. For the deberta (v1) model, they were implemented in #11387. 

Deberta v2 fast tokenizers are yet to be implemented.

### LINKED ISSUE #14712: DeBERTa V3 Fast Tokenizer
# 🚀 Feature request

Fast Tokenizer for DeBERTA-V3 and mDeBERTa-V3 

## Motivation

DeBERTa V3 is an improved version of DeBERTa. With the V3 version, the authors also released a multilingual model "mDeBERTa-base" that outperforms XLM-R-base. However, DeBERTa V3 currently lacks a FastTokenizer implementation which makes it impossible to use with some of the example scripts (They require a FastTokenizer). 
DeBERTa-V1 and DeBERTa-V2 both have a FastTokenizer implementation, it would be great to have one for DeBERTa-V3.

### PR DESCRIPTION
**Pull request by alcinos**
# What does this PR do?

Implements a fast tokenizer for deberta v2. Loosely based on #11387

Fixes #11529
Fixes #14712 

This is a draft as there are some failing tests (not super clear to me why atm, will have to investigate)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@LysandreJik

### CODE REVIEW

File: docs/source/main_classes/tokenizer.mdx [hunk 1]
```diff
@@ -18,9 +18,7 @@ Rust library [🤗 Tokenizers](https://github.com/huggingface/tokenizers). The "
 
 1. a significant speed-up in particular when doing batched tokenization and
 2. additional methods to map between the original string (character and words) and the token space (e.g. getting the
-   index of the token comprising a given character or the span of characters corresponding to a given token). Currently
-   no "Fast" implementation is available for the SentencePiece-based tokenizers (for T5, ALBERT, CamemBERT, XLM-RoBERTa
-   and XLNet models).
+   index of the token comprising a given character or the span of characters corresponding to a given token).
```
Comment (alcinos) [inline docs/source/main_classes/tokenizer.mdx:21]: Note: this change is not technically related to the PR, but I found this bit of the documentation to be outdated. If you want a separate PR, lmk.

**Comment by alcinos** [comment]
I noticed that while I was working on my PR, another was submitted for the same purpose: #14923.

**Comment by stefan-it** [comment]
Hey @alcinos ,

thanks for adding it!

I'm currently running comparisons between slow and fast tokenizer. Here are some mismatches between fast and slow.

I just run tokenization tests on `README.md` and `README_zh-hans.md` from official Transformers library, using this script:

```python
import sys

from transformers import DebertaV2Tokenizer, DebertaV2TokenizerFast

model_name = "microsoft/deberta-v2-xlarge"

slow_tokenizer = DebertaV2Tokenizer.from_pretrained(model_name)
fast_tokenizer = DebertaV2TokenizerFast.from_pretrained(model_name)

filename = sys.argv[1]

with open(filename, "rt") as f_p:
    for line in f_p:
        line = line.rstrip()

        if not line:
            continue

        slow_tokens = slow_tokenizer.tokenize(line)
        fast_tokens = fast_tokenizer.tokenize(line)

        if slow_tokens != fast_tokens:
            print("Tokenization mismatch:", line)
            print("Slow tokens:", slow_tokens)
            print("Fast tokens:", fast_tokens)
```

Here are some mismatches:

Original input: `* 🖼️ Images, for tasks like image classification, object detection, and segmentation.`
Slow tokens: `['▁*', '▁', '[UNK]', '️', '▁Images', ',', '▁for', '▁tasks', '▁like', '▁image', '▁classification', ',', '▁object', '▁detection', ',', '▁and', '▁segmentation', '.']`
Fast tokens: `['▁*', '▁', '🖼', '️', '▁Images', ',', '▁for', '▁tasks', '▁like', '▁image', '▁classification', ',', '▁object', '▁detection', ',', '▁and', '▁segmentation', '.']`

Another example on `README_zh-hans.md`:

Original input: `- 对教学和实践友好且低门槛`
Slow tokens: `['▁-', '▁', '对', '教', '学', '和', '实', '践', '友', '好', '且', '低', '门', '[UNK]']`
Fast tokens: `['▁-', '▁', '对', '教', '学', '和', '实', '践', '友', '好', '且', '低', '门', '槛']`

I then used the original DeBERTa tokenizer, and the mismatch is getting huge, e.g.:

Original input: `pages = "38--45"`
Tokens (Original Tokenizer): `['▁pages', '▁=', '▁"', '38', '-', '-', '45', '"']`
Tokens (Slow Tokenizer):     `['▁pages', '▁=', '▁"', '38', '-', '-45', '"']`
Tokens (Fast Tokenizer):     `['▁pages', '▁=', '▁"', '38', '-', '-45', '"']`

So there's also a mismatch between original DeBERTa tokenizer and the current one in Transformers.

**Review by SaulLu — COMMENTED** [review]
Thank you so much for working to resolve this issue @alcinos ! I see that you have already understood how our tokenizers were designated: that's really great! :smile: 

I took the liberty to leave a little comment on a particular point of this tokenizer, don't hesitate if you need more information about it.

By the way, I noticed that you said "This is a draft as there are some failing tests (not super clear to me why atm, will have to investigate)". Are all the tests that don't pass not clear to you, or only some? In any case, we can list together the tests that do not pass in this PR and discuss how to solve these problems. 

PS: I prefer to warn you that I will probably not be very active this week on github, as I'm on vacations.

File: src/transformers/models/deberta_v2/tokenization_deberta_v2_fast.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    pretrained_vocab_files_map = PRETRAINED_VOCAB_FILES_MAP
+    pretrained_init_configuration = PRETRAINED_INIT_CONFIGURATION
+    max_model_input_sizes = PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES
+    slow_tokenizer_class = DebertaV2Tokenizer
+
+    def __init__(
+        self,
+        vocab_file=None,
+        do_lower_case=False,
+        split_by_punct=False,
```
[the remark this answers is not in the document]
Comment (alcinos) [reply src/transformers/models/deberta_v2/tokenization_deberta_v2_fast.py:116]: Hello @SaulLu 
Thanks a lot for your review.

As for the lower_case arg, I followed Alberts’s tokenizer. As you mentioned, Albert has a modification to the normalizer in the converter:
https://github.com/huggingface/transformers/blob/501307b58bdc2db1b6a25271a3f60975130f1c6c/src/transformers/convert_slow_tokenizer.py#L502-L511
which I duplicated in my PR:
https://github.com/huggingface/transformers/blob/c6f2e1bbad3f673a4d1df62109e580f1c209de0e/src/transformers/convert_slow_tokenizer.py#L825-L826

Eyeballing the `init` method of `PreTrainedTokenizerFast` makes me believe the creation process always involves using the  said slow->fast conversion method, so that should be covered?


As for `split_by_punct` we could take the same approach and overload the pre-tokenizer method of the converter? Would a sequence [MetaSpace, Punct] do the trick? I’m a bit uncertain here since there doesn’t seem to be any other converter that seem to be dealing with punctuation splitting so maybe I’m understanding this wrong.

**Comment by alcinos** [comment]
@stefan-it Thanks for looking into this and providing the testcases.

It seems that the issues you are reporting are all related to unknown tokens? I don’t know the rust implementation well enough, is there any reason why the fast tokenizer would not respect the vocabulary?

**Comment by stefan-it** [comment]
Hey @alcinos I'm currently trying to figure it out. At the moment, I'm also getting no UNKs when using T5Tokenizer, but I'm investigating it.

**Comment by stefan-it** [comment]
Good news: when using `encode`, there's no mismatch between slow and fast tokenizer.

For slow tokenizer, this is happening here:

https://github.com/huggingface/transformers/blob/master/src/transformers/models/deberta_v2/tokenization_deberta_v2.py#L326-L330

Also when using normal T5 (Slow and fast) there are no UNKs when using the `tokenize` function (but `encode` shows that those subtokens are UNKs) so this is DeBERTa-specific.

**Comment by mingboiz** [comment]
@alcinos  I think the issue is regarding the tokenize function from `PreTrainedTokenizerFast` 

`encode_plus` returns the correct Dict[List] converted BatchEncoding of `Input-ids`. For both slow and fast tokenizers they return ```{'input_ids': [943, 250, 3, 28596, 7654, 6, 14, 2930, 72, 812, 8692, 6, 2328, 5563, 6, 7, 27235], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}```

https://github.com/huggingface/transformers/blob/501307b58bdc2db1b6a25271a3f60975130f1c6c/src/transformers/tokenization_utils_fast.py#L316-L317
 but the .tokens() at the end of the line causes the discrepancy here. 

Was able to fix the discrepancy in the testing script @stefan-it provided by override the tokenize function
```
    def tokenize(self, text: str, pair: Optional[str] = None, add_special_tokens: bool = False, **kwargs) -> List[str]:
        enc = self.encode_plus(text=text, text_pair=pair, add_special_tokens=add_special_tokens, **kwargs)
        return self.convert_ids_to_tokens(enc['input_ids'])
```

Both the other tests are still failing and I'm still trying to investigat

**Comment by alcinos** [comment]
Thanks @stefan-it and @mingboiz for looking into the tokenization issue. If I summarize the findings so far:

- Slow tokenizer replaces unknown tokens with "[UNK]" while fast tokenizer doesn’t
- This behavior seems specific to Deberta, as T5’s tokenizers don’t replace with "[UNK]"
- After encoding, the results are the same for slow and fast, meaning that the issue is probably minor
- @mingboiz found a way to have the fast tokenizer spit out the "[UNK]".

I’m not sure what the expected behavior should be, nor whether we should be concerned about this in the first place. Input from someone from HF would be appreciated :) (ping @SaulLu )


Aside from that, I pushed some fixes, more tests are passing.
Some feedback for the HF team on the issue I ran into:
In one of the common tests, the code looks for a "do_lower_case" attribute variable: 
https://github.com/huggingface/transformers/blob/10fd4fa1a6ca08b6bba5fed2db666a55239d717c/tests/test_tokenization_common.py#L626-L627
This is problematic in my opinion since:
- I didn’t see any mention that this attribute variable is required in the documentation I’ve come across (though I may have missed it)
- The code silently fails if the attribute is not present
- This argument itself is not used anywhere in the DebertaV2TokenizeFast class, hence I was not naturally inclined to add it as an attribute.

I would suggest one of the following change to make this more dev friendly:
- Make it a hard requirement that any tokenizer class must have this attribute, document it, and remove the silent fail in case it’s not found
- Additionally, IMHO this would be better suited as a an overridable getter method rather than a direct access to a private attribute


More tests are failing: one most likely has the same root cause as the issue raised by @stefan-it. The others seem to be failing because in some code paths the vocab_file is None, but it’s not clear to me why that happens, any help on that appreciated.

**Comment by mingboiz** [comment]
@alcinos I can't figure out a solution yet but the tests that are failing because of the missing vocab file is because in all of them legacy=False is being selected which only saves using the Rust Tokenizer these files: https://github.com/huggingface/transformers/blob/f80775df2b418716acce76d56826ed59183869b6/tests/test_tokenization_common.py#L3513

```
tokenizer_config.json',
special_tokens_map.json',
tokenizer.json'
```
without the `spm.model` vocab file being saved as this code chunk will run instead without using your save_vocabulary function:

https://github.com/huggingface/transformers/blob/f80775df2b418716acce76d56826ed59183869b6/src/transformers/tokenization_utils_fast.py#L578-L583

Debertav2Tokenizer didn't have this issue because its backend SPMTokenizer class provided its own `save_pretrained` method to save the `spm.model`, but I can't figure out why the AlbertTokenizerFast tests could work but the same tests fails here - I will see if I can fix this, but any help would be appreciated!

**Comment by SaulLu** [comment]
I have noted the ping! I'll come back to you as soon as possible on this subject because the choice to be made here is not obvious: you have highlighted that the `tokenize` method of `DebertaV2Tokenizer` does not behave in the same way as all the tokenizer methods of the fast tokenizers. At the moment it would make more sense to me to modify the `DebertaV2Tokenizer`'s `tokenize` method, but in general we don't really like to introduce backward incompatibility, so I'll need to discuss it with the maintainers :slightly_smiling_face:.

**Review by SaulLu — COMMENTED** [review]
Thanks again for all your work, I'm sorry I couldn't be very responsive last week.

So, to bounce around a few points: 
> I’m not sure what the expected behavior should be, nor whether we should be concerned about this in the first place. Input from someone from HF would be appreciated :)

@alicnos, @minboiz, One of the maintainers also shares my opinion, we think here that it would be better to modify the `tokenize` method of the slow version of the tokenizer so that it looks like the one of other tokenizers.

> Some feedback for the HF team on the issue I ran into:

@alcinos , thanks a lot for sharing your opinion, it's extremely useful to know the difficulties you encounter. 

Indeed, when you add a new tokenizer, an error like the one you mention with `do_lower_case` happens quickly because all the tests in `test_tokenization_commons.py` are not useful for all the tokenizers (and can be silently passed). My advice is to voluntarily introduce an error (`raise ValueError("to delete")` in the most used methods (like `encode`, `tokenize`, `__call__`) in order to see which tests are really executed and which are not. I then check that it is normal if some tests are not executed . 

Unfortunately, not all tokenizers in the library have a `do_lower_case` argument and that's why this is not a hard requirement.

> Additionally, IMHO this would be better suited as an overridable getter method rather than a direct access to a private attribute

@alcinos , I'm not sure what you are referring to here. What private attribute are you talking about? :slightly_smiling_face: 

**DebertaV2TokenizationTest.test_saving_tokenizer_trainer test fails**
I'll try to find out more about what's going on. From what I have seen this is due to the fact that on the first save, the key and the value ` 'tokenizer_file': None` are saved in the `tokenizer_config.json` (so they are not overridden afterwards because of [this line](https://github.com/huggingface/transformers/blob/v4.15.0/src/transformers/tokenization_utils_base.py#L1872)). What I need to check again is why this key and value are saved with deberta-v2 but not with albert.

**Comment by SaulLu** [comment]
@alcinos and @mingboiz , while investigating the `test_saving_tokenizer_trainer` test further, I noticed that the variable `VOCAB_FILES_NAMES`  did not specify the `"tokenizer_file"` value (and we need it for the fast version of the tokenizer). Moreover, for the test to fully succeed, the following lines must also be removed from the `tokenization_deberta_v2_fast.py` file.

```
        if not os.path.isfile(vocab_file):
            raise ValueError(
                f"Can't find a vocabulary file at path '{vocab_file}'. To load the vocabulary from a Google pretrained "
                "model use `tokenizer = AutoTokenizer.from_pretrained(PRETRAINED_MODEL_NAME)`"
            )
```
With these 2 changes, the test now pass :smile: !

I have opened a [PR](https://github.com/alcinos/transformers/pull/1/files) here to show you the changes that should be made to solve these problems. Feel free to merge it if you agree with it. :slightly_smiling_face:

**Comment by stefan-it** [comment]
@alcinos could you please have a look at the https://github.com/alcinos/transformers/pull/1 PR - I think it is ready then :hugs:

File: src/transformers/models/deberta_v2/tokenization_deberta_v2_fast.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+
+        if not os.path.isdir(save_directory):
+            logger.error(f"Vocabulary path ({save_directory}) should be a directory")
+            return
+        out_vocab_file = os.path.join(
+            save_directory, (filename_prefix + "-" if filename_prefix else "") + VOCAB_FILES_NAMES["vocab_file"]
+        )
+
+        if os.path.abspath(self.vocab_file) != os.path.abspath(out_vocab_file):
+            copyfile(self.vocab_file, out_vocab_file)
```
Comment (helpmefindaname) [inline src/transformers/models/deberta_v2/tokenization_deberta_v2_fast.py:245]: Can you please allow the tokenizer to be also saved, if the file it loaded from is removed?
(see [here](https://github.com/huggingface/transformers/commit/ade7371a41145c53fe90f66da7e5606f96068e98) for example `tokenization_albert.py` )

**Review by helpmefindaname — CHANGES_REQUESTED** [review]

File: src/transformers/models/deberta_v2/tokenization_deberta_v2_fast.py (hunk 3 above)
Comment (SaulLu) [reply src/transformers/models/deberta_v2/tokenization_deberta_v2_fast.py:245]: That's a really good point!

**Comment by SaulLu** [comment]
Hi @alcinos, thank you very much for your work, the addition seems to be near the end! Please let me know if you need help with any of it!

**Comment by SaulLu** [comment]
Hi @alcinos, thank you very much for your work, the addition seems to be near the end! Please let me know if you need help with any of it!

File: src/transformers/models/deberta_v2/tokenization_deberta_v2_fast.py (hunk 3 above)
Comment (mingboiz) [reply src/transformers/models/deberta_v2/tokenization_deberta_v2_fast.py:245]: Isn't this already supported in deberta-v2? specifically line 482, 483

https://github.com/huggingface/transformers/blob/5c2d8391984372f15a57c4be9a04d0effe037de9/src/transformers/models/deberta_v2/tokenization_deberta_v2.py#L476-L483

**Comment by SaulLu** [comment]
Finished in PR #15529

Thanks all again for the contribution :hugs:
